src/libraries/Common/src/System/Net/IPv4AddressHelper.Common.cs

```diff
@@ -176,62 +190,79 @@ internal static unsafe bool IsValidCanonical(char* name, int start, ref int end,
         // Return Invalid (-1) for failures.
         // If the address has less than three dots, only the rightmost section is assumed to contain the combined value for
         // the missing sections: 0xFF00FFFF == 0xFF.0x00.0xFF.0xFF == 0xFF.0xFFFF
-        internal static unsafe long ParseNonCanonical(char* name, int start, ref int end, bool notImplicitFile)
+        internal static long ParseNonCanonical<TChar>(in ReadOnlySpan<TChar> name, out int charsConsumed, bool notImplicitFile)
+            where TChar : unmanaged, IBinaryInteger<TChar>
         {
-            int numberBase = Decimal;
-            char ch;
-            long* parts = stackalloc long[4];
+            int numberBase = IPv4AddressHelper.Decimal;
+            Span<uint> parts = stackalloc uint[4];
             long currentValue = 0;
             bool atLeastOneChar = false;
 
             // Parse one dotted section at a time
             int dotCount = 0; // Limit 3
-            int current = start;
-            for (; current < end; current++)
+            int current;
+
+            Debug.Assert(typeof(TChar) == typeof(char) || typeof(TChar) == typeof(byte));
+
+            charsConsumed = 0;
+            for (current = 0; current < name.Length; current++)
             {
-                ch = name[current];
+                TChar ch = name[current];
+                int maxCharacterValue = '9';
                 currentValue = 0;
 
-                // Figure out what base this section is in
-                numberBase = Decimal;
-                if (ch == '0')
+                // Figure out what base this section is in, default to base 10
+                numberBase = IPv4AddressHelper.Decimal;
+                // A number starting with zero should be interpreted in base 8 / octal
+                // If the number starts with 0x, it should be interpreted in base 16 / hex
+                if (ch == TChar.CreateTruncating('0'))
                 {
-                    numberBase = Octal;
                     current++;
                     atLeastOneChar = true;
-                    if (current < end)
+                    if (current < name.Length)
                     {
                         ch = name[current];
-                        if (ch == 'x' || ch == 'X')
+
+                        if ((ch == TChar.CreateTruncating('x')) || (ch == TChar.CreateTruncating('X')))
```

Comment:
> How do you feel about an API proposal that extends the generic math interfaces specifically with

To what benefit? How would having those APIs change the code being written here?

Reply:
The code here would change to `TChar.ConvertToInt32Truncating(ch) == 'x'` for example, much as we did for cases like https://source.dot.net/#System.Runtime.Numerics/src/libraries/Common/src/System/Number.Parsing.Common.cs,363

It gives the same benefit that `IUtfChar<TSelf>` gave the code in corelib, it's a more direct path that can be better handled across all runtimes as it avoids the extra levels of generics and the many if/else type checks otherwise required to specialize (which runtimes like Mono don't support doing today and may not in the near future).

Reply:
Thanks. In that case personally I don't think it's worth it.